src/Internal/Payment/Storage.php

```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * Class Storage
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment;
+
+use WC_Order;
+use WCPay\Internal\Payment\State\InitialState;
+use WCPay\Internal\Payment\State\State;
+use WCPay\Internal\Service\PaymentMethodService;
+
+/**
+ * Storage for payments.
+ */
+class Storage {
+	/**
+	 * Key for storing data in order meta.
+	 *
+	 * @var string
+	 */
+	const ORDER_META_KEY = '_wcpay_payment_data';
+
+	/**
+	 * State factory.
+	 *
+	 * @var StateFactory
+	 */
+	private $state_factory;
+
+	/**
+	 * Payment methods service.
+	 *
+	 * @var PaymentMethodService
+	 */
+	private $payment_method_service;
+
+	/**
+	 * Controls whether storage is actually enabled.
+	 *
+	 * @var bool
+	 */
+	private $enabled = false;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param StateFactory         $state_factory          A factory for payment states.
+	 * @param PaymentMethodService $payment_method_service PM service.
+	 */
+	public function __construct(
+		StateFactory $state_factory,
+		PaymentMethodService $payment_method_service
+	) {
+		$this->state_factory          = $state_factory;
+		$this->payment_method_service = $payment_method_service;
+	}
+
+	/**
+	 * Enables payment storage.
+	 *
+	 * IMPORTANT: By default, this class does not touch the database.
+	 * Once enabled, it will, but this should only be used for tests
+	 * before the final storage method is determined.
+	 */
+	public function enable() {
+		$this->enabled = true;
+	}
+
+	/**
+	 * Loads or creates the payment object for an order,
+	 * returning the one true payment object.
+	 *
+	 * @param WC_Order $order Order that requires/has payment.
+	 * @return State          Current state of the order payment.
+	 */
+	public function get_order_payment( WC_Order $order ) {
+		$payment     = new Payment( $order );
+		$state_class = InitialState::class;
+
+		if ( $this->enabled ) {
+			$data = $order->get_meta( self::ORDER_META_KEY );
+			if ( is_array( $data ) && ! empty( $data ) ) {
+				$state_class = $data['state'];
```

Comment:
`$data['state']` is only read here. The idea is that while storing the payment, the class name of the state would be included, but it would not be stored within the payment otherwise, avoiding a link bank from `Payment` to `State`.

Reply:
> if we make all saving/reading go through one place and manage the context correctly, we should be fine.

That's the point of this class. Storage ended up being a secondary responsibility.